acq4/util/mies.py

```diff
@@ -0,0 +1,148 @@
+    """
+    sigDataReady = QtCore.Signal(object)
+    _sigFutureComplete = QtCore.Signal(object)
+    _bridge = None
+    ALLDATA = None
+    PEAKRES = 1
+    SSRES = 2
+
+    @classmethod
+    def getBridge(cls, useZMQ=False):
+        """Return a singleton MIESBridge instance.
+        """
+        # TODO: Handle switching between ZMQ and ActiveX?
+        if cls._bridge is None:
+            cls._bridge = MIES(useZMQ=useZMQ)
+        return cls._bridge
+
+    def __init__(self, useZMQ=False):
+        super(MIES, self).__init__(parent=None)
+        self.igor = IgorThread(useZMQ)
+        self.usingZMQ = useZMQ
+        self.currentData = None
+        self._exiting = False
+        self.windowName = 'ITC1600_Dev_0'
+        self._sigFutureComplete.connect(self.processUpdate)
+        self._initTPTime = None
+        self._lastTPTime = None
+        self._TPTimer = QtCore.QTimer()
+        self._TPTimer.setSingleShot(True)
+        self._TPTimer.timeout.connect(self.getMIESUpdate)
+        self.start()
+
+    def start(self):
+        self._exiting = False
+        if self.usingZMQ:
+            self.getMIESUpdate()
+
+    def getMIESUpdate(self):
+        if self.usingZMQ:
+            future = self.igor("FFI_ReturnTPValues")
+            future.add_done_callback(self._sigFutureComplete.emit)
+            self._TPTimer.start(5000) # by default recheck after 5 seconds, overridden if we get data
+        else:
+            raise RuntimeError("getMIESUpdate not supported in ActiveX")
+
+    def processUpdate(self, future):
+        if not self._exiting:
+            res = future.result()
+            if isinstance(res, IgorCallError):
+                # Test pulse isn't running, let's wait a little longer
+                nextCallWait = 200
+            else:
+                data = res[...,0] # dimension hack when return value suddenly changed
+                self.currentData = data
+                self._updateTPTimes(data)
+                self.sigDataReady.emit(data)
+                nextCallWait = 0
+            self._TPTimer.start(nextCallWait)
+
+    def _updateTPTimes(self, TPArray):
+        """Update globally tracked initial and end TP times"""
+        if self._initTPTime is None:
+            try:
+                self._initTPTime = TPArray[0,:][TPArray[0,:] > 0].min()
+            except ValueError:
+                pass
+        self._lastTPTime = TPArray[0,:].max()
+
+    def getTPRange(self):
+        if self._initTPTime is None:
+            return 0, 0
+        else:
+            return self._initTPTime, self._lastTPTime
+
+    def resetData(self):
+        self.currentData = None
+
+    def selectHeadstage(self, hs):
+        return self.setCtrl("slider_DataAcq_ActiveHeadstage", hs)
+
+    def setManualPressure(self, pressure):
+        return self.setCtrl("setvar_DataAcq_SSPressure", pressure)
+
+    def setApproach(self, hs):
+        windowName = '"{}"'.format(self.windowName)
```

Comment:
Why are you adding quotes here? This seems to be causing an error in mies:

    Assertion FAILED in function IsControlDisabled(...) MIES_GuiUtilities.ipf:862.
    Message: Non-existing control or window "ITC1600_Dev_0"